### ISSUE TITLE
apfs_fs.cpp: use of std::string(char *c_string) without maximum size

### ISSUE THREAD

**Issue by joachimmetz** [sleuthkit/sleuthkit#2519]
In https://github.com/sleuthkit/sleuthkit/blame/b9a79defddb1c203db2be39d47320192748f2230/tsk/fs/apfs_fs.cpp#L90 std::string() is directly called on uncontrolled input data. This leads to an unbounded OOB read seeing this call will look for a terminating 0-byte value (essentially running strlen() on the input data)

```
_name = std::string(xfield_data, ext.len);
```

Might be more secure, adding @jtsylve since he appears to have written the code and I'm not too sure about this code seeing all the recasting that is happening.

This also leads to another OOB read since the bound of xfield_data and ext.len is not checked.

**Comment by jtsylve** [comment]
You are correct, but with one caveat.  The length includes the null terminator, so the correct approach would be 

```cpp
_name = std::string(xfield_data, ext.len - 1);
```

**Comment by joachimmetz** [comment]
thx for the additional so the code also needs an `ext.len >= 1` test as well then

**Comment by joachimmetz** [comment]
and to double check `e.value.count()` contains the size of the buffer that contains the inode and extended attributes?

**Comment by joachimmetz** [comment]
@jtsylve PTAL if this makes sense to prevent multiple OOB reads from occurring https://github.com/sleuthkit/sleuthkit/pull/2520
